File: test/tunnel_config_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:httpku/models/tunnel_config.dart';

void main() {
  group('TunnelConfig Tests', () {
    test('Default values match expected setup', () {
      final config = TunnelConfig();
      expect(config.name, 'default');
      expect(config.remoteAddr, 'id1.jagoanip.my.id');
      expect(config.remotePort, 443);
      expect(config.type, TunnelType.http);
      expect(config.httpAddr, 'bisnis.udemy.com');
      expect(config.httpPort, 8080);
      expect(config.replaceHttpResponse, isTrue);
      expect(config.customHttpResponse, 'HTTP/1.1 200 Connection established');
      expect(config.customPayload, isTrue);
      expect(config.detectIpv4, isTrue);
    });

    test('Serialization and deserialization works correctly', () {
      final config = TunnelConfig(
        remoteAddr: 'test.example.com',
        remotePort: 22,
        type: TunnelType.ssl,
      );

      final jsonStr = config.toJson();
      final restored = TunnelConfig.fromJson(jsonStr);

      expect(restored.remoteAddr, 'test.example.com');
      expect(restored.remotePort, 22);
      expect(restored.type, TunnelType.ssl);
    });

    test('CopyWith updates fields properly', () {
      final config = TunnelConfig();
      final updated = config.copyWith(
        remoteAddr: 'new.domain.com',
        type: TunnelType.direct,
      );

      expect(updated.remoteAddr, 'new.domain.com');
      expect(updated.type, TunnelType.direct);
      expect(updated.remotePort, config.remotePort);
    });
  });
}

File: app/src/test/kotlin/com/sslh/sshl/app/TunnelConfigTest.kt
package com.sslh.sshl.app

import com.sslh.sshl.app.model.TunnelConfig
import com.sslh.sshl.app.model.TunnelType
import org.junit.Assert.*
import org.junit.Test

class TunnelConfigTest {

    @Test
    fun testDefaultConfig() {
        val config = TunnelConfig()
        assertEquals("default", config.name)
        assertEquals("id1.jagoanip.my.id", config.remoteAddr)
        assertEquals(443, config.remotePort)
        assertEquals(TunnelType.HTTP, config.type)
        assertEquals("bisnis.udemy.com", config.httpAddr)
        assertEquals(8080, config.httpPort)
        assertEquals("1.1.1.1", config.dnsServer)
        assertTrue(config.detectIpv4)
    }

    @Test
    fun testJsonSerialization() {
        val config = TunnelConfig(
            name = "test_config",
            remoteAddr = "1.2.3.4",
            remotePort = 22,
            type = TunnelType.DNS,
            dnsServer = "8.8.8.8"
        )
        val json = config.toJson()
        val deserialized = TunnelConfig.fromJson(json)

        assertEquals("test_config", deserialized.name)
        assertEquals("1.2.3.4", deserialized.remoteAddr)
        assertEquals(22, deserialized.remotePort)
        assertEquals(TunnelType.DNS, deserialized.type)
        assertEquals("8.8.8.8", deserialized.dnsServer)
    }

    @Test
    fun testTunnelTypeFromString() {
        assertEquals(TunnelType.DIRECT, TunnelType.fromString("direct"))
        assertEquals(TunnelType.HTTP, TunnelType.fromString("http"))
        assertEquals(TunnelType.SSL, TunnelType.fromString("ssl"))
        assertEquals(TunnelType.SOCKS, TunnelType.fromString("socks"))
        assertEquals(TunnelType.DNS, TunnelType.fromString("dns"))
        assertEquals(TunnelType.HAPROXY, TunnelType.fromString("haproxy"))
        assertEquals(TunnelType.HTTP, TunnelType.fromString("unknown"))
    }
}
